bare-metal-fractal/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+
+extern "C" {
+    fn js_logu(x: u32);
+    fn js_logf(x: f32);
+}
+
+// Memory Management
+#[repr(C)]
+pub struct TileBuffer {
+    w: u32,
+    h: u32,
+    buf: Vec<u32>,
+}
+
+// http://jakegoulding.com/rust-ffi-omnibus/objects/
+#[no_mangle]
+pub extern "C" fn alloc_tile(width: u32, height: u32) -> *mut TileBuffer {
+    let tile = TileBuffer {
+        w: width,
+        h: height,
+        buf: Vec::with_capacity((width * height) as usize),
+    };
+    Box::into_raw(Box::new(tile))
+}
+
+#[no_mangle]
+pub extern "C" fn get_buffer(tile_ptr: *mut TileBuffer) -> *mut u32 {
+    let tile = unsafe {
+        assert!(!tile_ptr.is_null());
+        &mut *tile_ptr
+    };
+
+    tile.buf.as_mut_ptr()
+}
+
+#[no_mangle]
+pub extern "C" fn free_tile(tile_ptr: *mut TileBuffer) {
+    if tile_ptr.is_null() {
+        return;
+    }
+    unsafe {
+        Box::from_raw(tile_ptr);
+    }
+}
+
+// Complex coordination
+// https://rustwasm.github.io/wasm-bindgen/examples/julia.html
+#[derive(Clone, Copy)]
+struct Complex {
+    re: f64,
+    im: f64,
+}
+
+impl Complex {
+    fn square(self) -> Complex {
+        let re = (self.re * self.re) - (self.im * self.im);
+        let im = 2.0 * self.re * self.im;
+        Complex { re, im }
+    }
+
+    fn norm(&self) -> f64 {
+        (self.re * self.re) + (self.im * self.im)
+    }
+}
+
+impl Add<Complex> for Complex {
+    type Output = Complex;
+
+    fn add(self, rhs: Complex) -> Complex {
+        Complex {
+            re: self.re + rhs.re,
+            im: self.im + rhs.im,
+        }
+    }
+}
+
+// Mandelbrot maths
+fn mandel_iter(max_iter: u64, c: Complex) -> u64 {
+    let mut z: Complex = c;
+
+    let mut iter = 1;
+
+    while (z.norm() <= 4.0) && (iter < max_iter) {
+        z = c + z.square();
+        iter = iter + 1;
+    }
+
+    if iter == max_iter {
+        0
+    } else {
+        iter
+    }
+}
+
+const fn rgb(r: u8, g: u8, b: u8) -> u32 {
+    255 << 24 | ((b as u32) << 16) | ((g as u32) << 8) | (r as u32)
+}
+
+fn rgbf(r: f32, g: f32, b: f32) -> u32 {
+    rgb((r * 255.0) as u8, (g * 255.0) as u8, (b * 255.0) as u8)
+}
+
+fn scale_value(value: f32, v_low: f32, v_high: f32, s_low: f32, s_high: f32) -> f32 {
+    (((value - v_low) / (v_high - v_low)) * (s_high - s_low)) + s_low
+}
+
+fn hsv_to_rgb(h: f32, s: f32, v: f32) -> u32 {
+    let i = (h * 6.0) as u8 as f32;
+    let f = h * 6.0 - i;
+    let p = v * (1.0 - s);
+    let q = v * (1.0 - f * s);
+    let t = v * (1.0 - (1.0 - f) * s);
+    if h < 0.16666666666 {
+        rgbf(v, t, p)
+    } else if h < 0.33333333333 {
+        rgbf(q, v, p)
+    } else if h < 0.5 {
+        rgbf(p, v, t)
+    } else if h < 0.66666666666 {
+        rgbf(p, q, v)
+    } else if h < 0.83333333333 {
+        rgbf(t, p, v)
+    } else {
+        rgbf(v, p, q)
+    }
+}
+
+static BOTTOM: u32 = rgb(0, 0, 0);
+
+// golfing to do here...
+fn build_palette(size: usize) -> Vec<u32> {
+    let mut palette = Vec::with_capacity(size);
+
```

Comment:
Does something like this work?

```rust
(0..size).map(|i| {
  let m = i as f32 / size as f32;
  if m < 0.5 {
     // same
  } else {
    // same
  }
}).collect();
```

Reply:
Alternatively, since we go from i in 0 to size, won't this essentially always split the array into two equal sizes with one half being blue and the other half being orange? In that case I think we can get rid of the branch entirely by writing two iterators and then returning the concatenated and collected version.

Something like 
```rust
let blue_half = (i..(size/2 | 0)).map( let h = ... );
let orange_half = (((size/2 + 1) | 0)..size).map ( let h = ... );
blue_half.chain(orange_half).collect()
```

Reply:
Totally rewrote the palette generator to use a cycle of defined gradients and shrunk the code a few hundred bytes
...
and then some unscrutable side effect doubled it. eh???